File: haskell/tests/LexerTest.hs
{-# LANGUAGE OverloadedStrings #-}
module Main where

import Subleq.Lexer
import Data.Aeson
import qualified Data.ByteString.Lazy as B
import qualified Data.Map as Map
import System.Exit (exitFailure)
import Data.Text (Text)
import qualified Data.Text as T

data LexerTestCase = LexerTestCase {
    input :: String,
    expected :: [TokenJSON]
} deriving (Show)

data TokenJSON = TokenJSON {
    tjType :: String,
    tjValue :: String
} deriving (Show)

instance FromJSON LexerTestCase where
    parseJSON = withObject "LexerTestCase" $ \v -> LexerTestCase
        <$> v .: "input"
        <*> v .: "expected"

instance FromJSON TokenJSON where
    parseJSON = withObject "TokenJSON" $ \v -> TokenJSON
        <$> v .: "type"
        <*> v .: "value"

matchToken :: Token -> TokenJSON -> Bool
matchToken (Token t v) (TokenJSON ejT ejV) =
    show t == ejT && v == ejV

runTest :: (String, LexerTestCase) -> IO Bool
runTest (name, tc) = do
    case tokenize (input tc) of
        Left err -> do
            putStrLn $ "Lexer error in " ++ name ++ ": " ++ show err
            return False
        Right output -> do
            let expectedTokens = expected tc
            if length output /= length expectedTokens
                then do
                    putStrLn $ "Token count mismatch in " ++ name ++ ": Expected " ++ show (length expectedTokens) ++ ", got " ++ show (length output)
                    return False
                else do
                    let matches = zipWith matchToken output expectedTokens
                    if all id matches
                        then return True
                        else do
                            putStrLn $ "Token mismatch in " ++ name
                            return False

main :: IO ()
main = do
    content <- B.readFile "tests/assembler/lexer_data.json"
    case decode content :: Maybe (Map.Map String LexerTestCase) of
        Nothing -> do
            putStrLn "Failed to parse lexer_data.json"
            exitFailure
        Just testData -> do
            results <- mapM runTest (Map.toList testData)
            if all id results
                then putStrLn "Lexer Tests: PASS"
                else do
                    putStrLn "Lexer Tests: FAIL"
                    exitFailure

File: cpp/tests/assembler/lexer_test.cpp
#include "../../lexer.hpp"
#include "../tester.hpp"
#include <fstream>

bool lexer_test_func(const std::string& test_name, const json& data) {
    std::string input = data["input"];
    std::vector<Token> output;
    try {
        output = tokenize(input);
    } catch (const std::exception& e) {
        std::cerr << "Lexer error in " << test_name << ": " << e.what() << std::endl;
        return false;
    }

    auto expected = data["expected"];
    if (output.size() != expected.size()) {
        std::cout << "Token count mismatch in " << test_name << ": Expected " << expected.size() << ", got " << output.size() << std::endl;
        return false;
    }

    for (size_t i = 0; i < output.size(); i++) {
        if (output[i].type != expected[i]["type"] || output[i].value != expected[i]["value"]) {
            std::cout << "Mismatch at index " << i << " in " << test_name << ": Expected " << expected[i]["type"] << ":" << expected[i]["value"] << ", got " << output[i].type << ":" << output[i].value << std::endl;
            return false;
        }
    }

    return true;
}

int main() {
    std::string data_path = "tests/assembler/lexer_data.json";
    std::ifstream file(data_path);
    if (!file.is_open()) {
        std::cerr << "Could not open data file: " << data_path << std::endl;
        return 1;
    }

    json test_data;
    file >> test_data;

    tester::run<json>("Lexer Tests", test_data, lexer_test_func);

    return 0;
}
